### PR TITLE
Stakeholder maintenance automation

### PR DESCRIPTION
**Pull request by wpschaub**
The first, of hopefully many, automation posts  introducing one of many of our common engineetring syetem automations.

### CODE REVIEW

**Comment by wpschaub** [comment]
@wsbctechnicalblog it is not yet ready for review, which is why it is a DRAFT PR. I should be able to finish it by Tuesday morning, so that we can have our first automation (working group) post :)

**Comment by wpschaub** [comment]
@wsbctechnicalblog as soon as @erwsch1 and @604kev are OK with the post, you can pull the trigger once you are OK with it.

**Review by wsbctechnicalblog — COMMENTED** [review]
Great start of the automation series! One question: you have proposed 5 stages of delivery:

- Ship value! – Increase the delivery of value to customers. 
- Ship value faster! – Shorten the delivery cycle. 
- Ship the right value faster! – Monitor, learn, adapt, and pivot.     
- Ship the right and better value faster! – Improve quality! 
- Ship the right and better value faster while reducing cost and efficiency! – Reduce cost and simplify through automation

Why not simply say:

"You must find ways to ship the right and better value faster while reducing cost and efficiency! – Reduce cost and simplify through automation."

Any reason why you are advising engineers to do it gradually?

**Review by erwsch1 — DISMISSED** [review]

**Comment by wpschaub** [comment]
> Great start of the automation series! One question: you have proposed 5 stages of delivery:
> 
> * Ship value! – Increase the delivery of value to customers.
> * Ship value faster! – Shorten the delivery cycle.
> * Ship the right value faster! – Monitor, learn, adapt, and pivot.
> * Ship the right and better value faster! – Improve quality!
> * Ship the right and better value faster while reducing cost and efficiency! – Reduce cost and simplify through automation
> 
> Why not simply say:
> 
> "You must find ways to ship the right and better value faster while reducing cost and efficiency! – Reduce cost and simplify through automation."
> 
> Any reason why you are advising engineers to do it gradually?

@wsbctechnicalblog it is an extract from a storyline from our book and needs to be taken within its context. To be honest, I have seen too many failed attempts where engineers jumped straight into automation without understanding the process (what they are automating and why). Evolving into automation is what has worked for me and those I influenced.

For example, in the automation working group the engineers jumped straight into the tooling. PowerShell or Bash?!? We need to focus on the WHAT and WHY first, for example the 5-min to hello world automation. Choosing the hammer (product) will be the easy part.

Just think of the building code. We have amazing automation in terms of PowerShell scripts developed by PM, however, to this day we have not enabled it in the automated building code. Too complex? Too ambitious?  No bandwidth? These are the worrying use cases which sway me away from the "automate everything" now guidance.

That said, my guidance of Ship...Ship...Ship does not specify a time. It can happen in short iterations, but at a pace where we know WHAT we are automation, WHY we are automating it, and HOW we are automating it.

Bottom line, I will stick to my guidance, because I know it is feasible and practical, and delivers quality.

**Comment by wsbctechnicalblog** [comment]
> > Great start of the automation series! One question: you have proposed 5 stages of delivery:
> > 
> > * Ship value! – Increase the delivery of value to customers.
> > * Ship value faster! – Shorten the delivery cycle.
> > * Ship the right value faster! – Monitor, learn, adapt, and pivot.
> > * Ship the right and better value faster! – Improve quality!
> > * Ship the right and better value faster while reducing cost and efficiency! – Reduce cost and simplify through automation
> > 
> > Why not simply say:
> > "You must find ways to ship the right and better value faster while reducing cost and efficiency! – Reduce cost and simplify through automation."
> > Any reason why you are advising engineers to do it gradually?
> 
> @wsbctechnicalblog it is an extract from a storyline from our book and needs to be taken within its context. To be honest, I have seen too many failed attempts where engineers jumped straight into automation without understanding the process (what they are automating and why). Evolving into automation is what has worked for me and those I influenced.
> 
> For example, in the automation working group the engineers jumped straight into the tooling. PowerShell or Bash?!? We need to focus on the WHAT and WHY first, for example the 5-min to hello world automation. Choosing the hammer (product) will be the easy part.
> 
> Just think of the building code. We have amazing automation in terms of PowerShell scripts developed by PM, however, to this day we have not enabled it in the automated building code. Too complex? Too ambitious? No bandwidth? These are the worrying use cases which sway me away from the "automate everything" now guidance.
> 
> That said, my guidance of Ship...Ship...Ship does not specify a time. It can happen in short iterations, but at a pace where we know WHAT we are automation, WHY we are automating it, and HOW we are automating it.
> 
> Bottom line, I will stick to my guidance, because I know it is feasible and practical, and delivers quality.

I did not mean "automate everything now". I meant "automate just one tiny, super easy little thing (for example, automate sending an email with code duplication report), but do it by finding ways to ship the right and better value faster while reducing cost and efficiency! – Reduce cost and simplify through automation. Rinse, repeat" 

No need to specify intermediate steps: first ship it slow, then ship it faster, etc. I don't see the need to avoid trying to go straight to the fastest possible, most efficient ways to automate.

**Comment by wpschaub** [comment]
@alexbunardzic I am not an advocate to automate everything or to start with automation. Automate what is automatable, where automation adds value, and let automation opportunities arise through repetition, rather than looking for stuff to automate. At the end we are saying the same (ODDA, rinse & repeat), I am just more pedantic on what to automate. I am over cautious after seeing too much waste of automation that adds no value, is never used, or far too complex to maintain. Also, I have not recommended stages of delivery, but a mindset - again, you need to read it within the context of its source as referenced.

**Comment by wsbctechnicalblog** [comment]
Makes sense @wpschaub. My approach is more radical, but life would be super boring if we all had the same approach ;)

**Review by wsbctechnicalblog — APPROVED** [review]

**Comment by wpschaub** [comment]
Now to get @604kev thumbs up, so that we can pull the trigger on this post.

**Comment by wpschaub** [comment]
@wsbctechnicalblog you can pull the trigger!
image

PS: The drop-the-mic automation script will be my next big post :)

**Review by littleswancode — APPROVED** [review]
Approved by DevSecOps!
